### PR TITLE
Avoid gather when copying strings view from start of strings column

### LINKED ISSUE #6803: [BUG] String column copy performs a full gather
**Describe the bug**
While analyzing an Nsight trace of a Spark query, I noticed 10ms being spent on a filter.  Digging deeper, the filter time was spent in a `cudf::table` copy constructor, with most of the time being spent copying one string column within the table.

Spark queries often filter input data to remove nulls before further processing, and often nothing is filtered.  `cudf::apply_boolean_mask`, used to implement the Spark filter, has short-circuit logic to avoid a full gather and instead copy-constructs the output table if it realizes nothing will be filtered.  However copy-constructing a string column performs a full gather which can be much more expensive than simply copying the input column buffers.

In this specific case, the string column consists of 178200 rows and each row is around 200 characters.  The string copy constructor took almost 10 milliseconds on the GPU despite the column containing approximately 30MB of device memory.

**Steps/Code to reproduce bug**
Perform a copy of a string column where there are many strings that are relatively long (e.g.: 200+ characters per row).  I've attached a gzipped Parquet file containing the string column from the specific filter case mentioned above for reference.

[filtertest.parquet.gz](https://github.com/rapidsai/cudf/files/5563552/filtertest.parquet.gz)

**Expected behavior**
A copy of a strings column view when the view starts at base offset 0 should be performed at device memory speed, copy the underlying buffers rather than performing the more complicated gather computation (which also requires synchronizing and a DtoH transfer).

### PR DESCRIPTION
**Pull request by jlowe**
Fixes #6803 

This optimizes string slice copying in the case where the string column view starts at offset 0.  In that case the offset values do not need to be modified, and all the column buffers can be copied.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/src/strings/copying/copying.cu [hunk 1]
```diff
@@ -42,6 +44,38 @@ std::unique_ptr<cudf::column> copy_slice(strings_column_view const& strings,
   if (end < 0 || end > strings_count) end = strings_count;
   CUDF_EXPECTS(((start >= 0) && (start < end)), "Invalid start parameter value.");
   strings_count = cudf::util::round_up_safe<size_type>((end - start), step);
+  if (start == 0 && strings.offset() == 0) {
```
Comment (davidwendt) [inline cpp/src/strings/copying/copying.cu:47]: This does not account for the `step` or the `end` parameters.

File: cpp/src/strings/copying/copying.cu [hunk 2]
```diff
@@ -42,6 +44,38 @@ std::unique_ptr<cudf::column> copy_slice(strings_column_view const& strings,
...
+    // slice starts at the beginning, so do not need a gather to adjust offsets
+    auto offsets_column = make_numeric_column(
```
Comment (davidwendt) [inline cpp/src/strings/copying/copying.cu:49]: The `gather` call is really only necessary for `step > 1`.
This whole thing probably can be simplified with something like the following:
```
if( start==0 && strings.offset()==0 && step==1 ) {
  auto offsets_column = std::make_unique<cudf::column>( 
     cudf::slice(strings.offsets(),{0,strings_count+1}).front()),
     stream, mr );
  auto data_size = cudf::detail::get_value<int32_t>(offsets_column->view(), strings_count, stream);
  auto chars_column = std::make_unique<cudf::column>(
     cudf::slice(strings.chars(),{0,data_size}).front(),
     stream, mr );
  return make_strings_column(strings_count,
                             std::move(offsets_column),
                             std::move(chars_column),
                             UNKNOWN_NULL_COUNT,
                             copy_bitmask(strings.null_mask(), 0, strings_count, stream, mr),
                             stream,
                             mr);
}
// do the gather instead
```

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/src/strings/copying/copying.cu (hunk 1 above)
Comment (jlowe) [reply cpp/src/strings/copying/copying.cu]: Nice catch!  This is fixed.

File: cpp/src/strings/copying/copying.cu (hunk 2 above)
Comment (jlowe) [reply cpp/src/strings/copying/copying.cu]: Thanks for the suggestion, this is much cleaner!  Updated.

**Review by davidwendt — APPROVED** [review]
Looks good. Can you provide some measurements on how much this improved performance?

**Comment by jlowe** [comment]
> Can you provide some measurements on how much this improved performance?

The speedup will be dependent upon the shape of the string column.  I used the data example from #6803 (string column with 178200 rows, each row has approximately 200 characters) and called `apply_boolean_mask` with a mask that does not filter anything, triggering the table copy constructor short-circuit.  The original table copy constructor for this one-column table takes approximately 10.46 msec and is synchronous with the GPU as shown in this Nsight profile:

image

With the new string slice optimization for start offset == 0 and step == 1, the same table copy constructor takes 116.354 usec on the CPU and runs asynchronously with the GPU as shown in this Nsight profile:

image

Including the time taken to complete the copies on the GPU for a fair comparison with the original table constructor, it takes approximately 180 usec from the start of the table constructor on the CPU to the last corresponding GPU operation to complete.  For this input data, performing the string column copy is approximately 58X faster than before.

Micro benchmarking aside, I measured over a full second faster wallclock time on a query like TPCx-BB query 21 at scale factor 100 using Spark default configs with this change applied (16.7 seconds vs 17.9 seconds).

File: cpp/tests/replace/replace_nulls_tests.cpp [hunk 3]
```diff
@@ -158,7 +158,7 @@ TEST_F(ReplaceNullsStringsTest, ReplaceNone)
   std::unique_ptr<cudf::column> result;
   ASSERT_NO_THROW(result = cudf::replace_nulls(input_w, replacement_w, mr()));
 
-  CUDF_TEST_EXPECT_COLUMNS_EQUAL(*result, expected_w);
+  CUDF_TEST_EXPECT_COLUMNS_EQUIVALENT(*result, expected_w);
```
Comment (harrism) [inline cpp/tests/replace/replace_nulls_tests.cpp:161]: Why this change?

**Review by harrism — APPROVED** [review]
Looks good, but the change to `replace_nulls_tests` stands out as unrelated to the optimization. Can you explain that one?

File: cpp/tests/replace/replace_nulls_tests.cpp (hunk 3 above)
Comment (jlowe) [reply cpp/tests/replace/replace_nulls_tests.cpp:161]: The test was failing because it expected there to be no null mask on the result column, but after the optimization it can be there.  That's why I changed it to check for equivalence testing.

When copying the string with a full gather, the null count is computed and the null mask can optimized out, but when copying the string view via just buffer copies, it does not compute the null count and instead copies the bitmask range of the input if the input has a mask.  In this test case the input does have a mask even though the null count is zero, so the output also has a mask.

If we want to optimize out the null mask, the code can compute the null count on the input string view and only copy the incoming view's null mask if there are any nulls.  That will change the code to synchronize with the GPU, but it could save a validity mask alloc and copy in the case where there are no nulls.  If that's desired I can update, and this test change will be removed.

Comment (harrism) [reply cpp/tests/replace/replace_nulls_tests.cpp:161]: I don't have an opinion on this.

Comment (jlowe) [reply cpp/tests/replace/replace_nulls_tests.cpp:161]: Yeah I similarly don't have a strong opinion here.  In some cases it will be better to avoid the null count, and sometimes it will be worth the memory savings.

Is this ready to go as-is, or do we need to find more opinions on this?
